applications/crossbar/priv/couchdb/account/devices.json

```diff
@@ -20,6 +20,24 @@
                 "}"
             ]
         },
+        "listing_by_external_cidnum": {
+            "map": [
+                "function (doc) {",
+                "  if (doc.pvt_type != 'device' || doc.pvt_deleted) return;",
+                "  if(doc.caller_id.external.number)",
+                "    emit(doc.caller_id.external.number, {'id': doc._id})",
+                "}"
+            ]
+        },
+        "listing_by_internal_cidnum": {
+            "map": [
+                "function (doc) {",
+                "  if (doc.pvt_type != 'device' || doc.pvt_deleted) return;",
+                "  if(doc.caller_id.internal.number)",
```

Comment:
Also, we do support "custom" caller IDs. I wonder if instead of having a view for internal and a view for external, if you just have one view that emits a composite key like:
```js
function(doc) {
  if (typeof doc.caller_id === "undefined"); return;

  for(var type in doc.caller_id) {
    emit([doc[type].number, type], doc.pvt_type);
  }
}
```
Untested, but the idea is you would get rows like:
```erlang
{["+1234567890", "external"], "user"}
,{["1005", "internal"], "device"}
 ,...
```

Then the query in the callflow becomes:
```erlang
{key, [kapps_call:callee_id_number(Call)
      | case kz_json:get_ne_binary_value(<<"caller_id_type">>, Data) of 'undefined' -> []; Type -> [Type] end %% this should be a function
      ]
}
```

Again, all untested but it feels like it could clean up a lot of duplicate code and allow people to create callflows triggering on custom caller IDs depending on how the call entered the system.

Then you can iterate over the results to create the endpoints (since you have the types of the document in the "value" you can filter there based on Data too).

Reply:
To add, the view can go in "attributes.json" design doc.